docs/hacking.rst

```diff
@@ -2,273 +2,256 @@
+  mc.addMessages(brandTranslations);
+  mc.addMessages(mainTranslations);
+
+  // An array for collection formatting errors.
+  const errors = [];
 
   const str1 = mc.format('hello-world', {
-    userName: 'John'
-  });
+    username: 'John'
+  }, errors);
 
-  // str1 === "Hello John! welcome to Loki"
+  // str1 === "Hello John! Welcome to Loki"
+  // errors.length == 0
 
-Localization API
-===================
+  const str2 = mc.format('hello-world', null, errors);
 
-Second layer has two fundamental features over the first one.
-It interacts with environment by using I/O to load resources and it is multi-lignual
-with language fallbacking.
+  // str1 === "Hello username! Welcome to Loki"
+  // errors.length == 1
+  // errors[0] is a <ReferenceError: Unknown external: username>
 
-Due to those two differences, it's fully asynchronous.
 
-ResourceBundle
---------------
+The Localization API
+====================
 
-A bundle is a single-locale object that holds collection of resource sources
-necessary for a given ``Localization`` object and corresponding ``MessageContext`` object.
+The ``Localization`` class represents the main functionality of l20n.js.  It 
+can retrieve formatted translations in accordance to the user's language 
+preferences, resorting to fallback languages in case of errors.  It also 
+interacts with the environment via I/O in order to load resources.  It is the 
+work-horse of l20n.js.  
 
-``ResourceBundle`` API:
+In Web-like runtimes the ``Localization`` API is fully asynchronous.  This 
+allows for loading fallback resources lazily in case of errors in the first 
+language.
 
-:lang:
-    Language tag associated with the bundle
+The ``Localization`` class can be thought of as a multi-lingual, asynchronous 
+wrapper on top of multiple ``MessageContext`` instances which handles lazy-IO 
+and the fallback strategy.
 
-:loaded:
-    Either ``false`` or a promise that returns a list of resources for the bundle.
+Internally, ``Localization`` instances use an array of runtime-specific 
+``ResourceBundle`` objects which represent information about localization 
+resources: in particular, how to load them when they're needed, and where from.  
+This array is the *de facto* fallback chain that will be used in case of 
+fallback.  For each fetched ``ResourceBundle`` the ``Localization`` creates 
+a corresponding ``MessageContext`` instance which it will use to format 
+translations to the bundle's language.
 
-:resIds:
-    List of resources for the bundle.
+The ``Localization`` constructor takes two runtime-specific functions as 
+arguments, ``requestBundles`` and ``createContext``. This allows the user to 
+provide different ways to build bundles, use different IO and language 
+negotiation methods, as well as pass custom functions to the ``MessageContext``
+constructor.  The ``Localization`` class itself doesn't know anything about the
+environment it runs in and doesn't provide any special methods to localize it. 
```

Comment:
I think I'll just mention that `Localization` is env-agnostic.  The first sentence in the next section explains this better anyways.